bootstrap/controllers/upgrade_controller.go

```diff
@@ -287,6 +291,13 @@ func (r *InPlaceUpgradeReconciler) handleUpgradeDone(ctx context.Context, scope
 func (r *InPlaceUpgradeReconciler) handleUpgradeFailed(ctx context.Context, scope *UpgradeScope) (reconcile.Result, error) {
 	mAnnotations := scope.Machine.GetAnnotations()
 
+	// NOTE(Hue): We don't remove the `LastFailedAttemptAt` annotation here
+	// because we want to know if the upgrade failed at some point in the `MachineDeploymentReconciler`.
+	// This function triggers a retry by removing the `Status` and `ChangeID` annotations,
+	// but the `LastFailedAttemptAt` lets us descriminiate between a retry and a fresh upgrade.
+	// Overall, we don't remove the `LastFailedAttemptAt` annotation in the `InPlaceUpgradeReconciler`.
+	// That's the responsibility of the `MachineDeploymentReconciler`.
```

Comment:
Mote: I'm not super happy about this solution. It causes an implicit dependency between two controllers but I don't really have an alternative solution so I'm fine with it for now.

Reply:
I understand. This coupling is indeed not desirable. But given the other solutions that we discussed, I feel like this is an acceptable workaround so far.